### PR TITLE
more vaapi build changes

### PR DESCRIPTION
**Pull request by Gusar321**
I noticed that my -Dwayland=disable build would still link to libva-wayland, this PR fixes that. There's also another drm_egl->drm change that I missed in my previous PR.

### CODE REVIEW

**Comment by kasper93** [comment]
I peeked at this code and considering this pattern:
```
vaapi = dependency('libva', version: '>= 1.1.0', required: get_option('vaapi'))
features += {'vaapi': vaapi.found() and features['libdl'] and (features['x11'] or
             features['wayland'] or egl_drm.allowed())}
             features['wayland'] or features['drm'])}
```
I don't like how we disable a _feature_, while not respecting `get_option('vaapi')`. `-Dvaapi=enabled` would only error out if `libva` is not found, but still `vaapi` _feature_ could be disabled as a result of other conditions.

Something like this, would properly respect the `vaapi`  option:
```
libva = dependency('libva', version: '>= 1.1.0', required: get_option('vaapi'))
vaapi.require(libva.found() and features['libdl'] and (features['x11'] or
              features['wayland'] or egl_drm.allowed())}
              features['wayland'] or features['drm'])) # optionally with nice error message
features += {'vaapi': vaapi.allowed()}
```

Just a food for thought, it requires more refactoring to "fix" things like that.

There are more suspicious patters, like requiring feature twice and so on. I'm sure it kinda works, but at lest for me it is structured in a way that is hard to follow dependency tree.

**Comment by Gusar321** [comment]
Yeah, all the vaapi handling still looks very weird to me even after my changes, but I don't know enough about meson to know what the proper approach is.

Basically, you need libva and then at least one of libva-x11, libva-drm, libva-wayland, otherwise you end up with "vaapi support" that doesn't actually work cos there's no way to create a VA display. Then there's two ways to render the vaapi video - EGL or vulkan/libplacebo. Or, in the case of X11, there's also vo_vaapi, No clue how to properly describe all that in meson talk.

**Comment by kasper93** [comment]
Let's just merge this as is. And we can improve on meson later with more in-depth look if someone get too much time on thier hand to look at it. Now that waf is removed I believe we have more freedom to change the meson build system, without worrying about waf parity.

**Comment by Gusar321** [comment]
Apparently I have too much time on my hands :) I refactored things based on your example.

**Comment by Dudemanguy** [comment]
Looking better thanks. I think `vaapi_wayland` also needs to require `features['wayland']` in the library check though? Same for x11 and drm.

File: meson.build [hunk 1]
```diff
@@ -1394,36 +1394,38 @@ if features['rpi-mmal']
...
-if features['vaapi']
-    dependencies += vaapi
-    sources += files('video/filter/vf_vavpp.c',
-                     'video/vaapi.c')
-endif
-
-vaapi_drm = dependency('libva-drm', version: '>= 1.1.0', required: get_option('vaapi-drm').require(features['vaapi']))
-features += {'vaapi-drm': features['vaapi'] and vaapi_drm.found()}
+vaapi_drm = dependency('libva-drm', version: '>= 1.1.0', required: get_option('vaapi-drm').require(libva.found()))
+features += {'vaapi-drm': features['drm'] and vaapi_drm.found()}
```
Comment (kasper93) [inline meson.build:1400]: I would also move `features['drm'] ` into `require()` above. But then you need to check twice, because `required: false` would not make `vaapi_drm` not found.

So basing it on the option would probably work, similar to previous
```
libva_drm = dependency('libva-drm', version: '>= 1.1.0', required: get_option('vaapi-drm'))
vaapi_drm = get_option('vaapi-drm').require(libva.found() and features['drm'] and libva_drm.found())
features += {'vaapi-drm': vaapi_drm}
```

And this made me think, we could even optimize configure to not search for libraries that would not be used. For example

```
vaapi_drm = get_option('vaapi-drm').require(libva.found() and features['drm'])
if vaapi_drm
    libva_drm = dependency('libva-drm', version: '>= 1.1.0')
    features += {'vaapi-drm': libva_drm.found()}
    if features['vaapi-drm'] # this conditional may not be required, I think by default meson doesn't return disabler()
        dependencies += libva_drm
    endif
else
       features += {'vaapi-drm': false}
endif
```

but I really dislike this nested conditional and the need to have false branch.

I'm also not meson master, @Dudemanguy may have better ideas. Looking at our `feature` array, I have an idea how to make it easier to handle. Might send RFC later.

Comment (Dudemanguy) [reply meson.build:1400]: `required: get_option('vaapi-drm').require(libva.found() and features['drm'])` skips trying to find libva-drm if you disable drm with `-Ddrm=disabled` so I think that change solves the issue or did you mean something else?

Comment (kasper93) [reply meson.build:1400]: I don't think so
```
dependency('libva-drm', version: '>= 1.1.0', required: false)
```
will still try to find the library, it will not fail if not found, but still does a lookup. It is not a problem, because configuration is fast anyway, just an optimization.

Comment (kasper93) [reply meson.build:1400]: Ah, my bad. I see now that `required` also takes a `feature`, not only `bool`, so `disabled` state is propagated and indeed search for lib would be skipped.

Than it is indeed, enough to put `features['drm']` into `require`.

**Comment by Gusar321** [comment]
You mean change
```
vaapi_wayland = dependency('libva-wayland', version: '>= 1.1.0', required: get_option('vaapi-wayland').require(libva.found()))
features += {'vaapi-wayland': features['wayland'] and vaapi_wayland.found()}
```
to
```
vaapi_wayland = dependency('libva-wayland', version: '>= 1.1.0', required: get_option('vaapi-wayland').require(features['wayland'] and libva.found()))
features += {'vaapi-wayland': vaapi_wayland.found()}
```
and similar for the other checks?

**Comment by Gusar321** [comment]
Ok, moved the feature[] stuff into the require checks.

**Comment by kasper93** [comment]
Looks nice to me now.

File: meson.build [hunk 2]
```diff
@@ -1394,36 +1394,38 @@ if features['rpi-mmal']
...
 
-if features['vaapi']
-    dependencies += vaapi
-    sources += files('video/filter/vf_vavpp.c',
-                     'video/vaapi.c')
-endif
-
-vaapi_drm = dependency('libva-drm', version: '>= 1.1.0', required: get_option('vaapi-drm').require(features['vaapi']))
-features += {'vaapi-drm': features['vaapi'] and vaapi_drm.found()}
+vaapi_drm = dependency('libva-drm', version: '>= 1.1.0', required: get_option('vaapi-drm').require(libva.found() and features['drm']))
```
Comment (Dudemanguy) [inline meson.build:1399]: Sorry just one last style thing. These lines are all pretty long could you break them into two like this?
```suggestion
vaapi_drm = dependency('libva-drm', version: '>= 1.1.0',
                                          required: get_option('vaapi-drm').require(libva.found() and features['drm']))
```

File: meson.build [hunk 3]
```diff
@@ -1394,36 +1394,38 @@ if features['rpi-mmal']
...
+features += {'vaapi-drm': vaapi_drm.found()}
 if features['vaapi-drm']
     dependencies += vaapi_drm
 endif
 
-vaapi_wayland = dependency('libva-wayland', version: '>= 1.1.0', required: get_option('vaapi-wayland').require(features['vaapi']))
-features += {'vaapi-wayland': features['vaapi'] and vaapi_wayland.found()}
+vaapi_wayland = dependency('libva-wayland', version: '>= 1.1.0', required: get_option('vaapi-wayland').require(libva.found() and features['wayland']))
```
Comment (Dudemanguy) [inline meson.build:1405]:
```suggestion
vaapi_wayland = dependency('libva-wayland', version: '>= 1.1.0',
                                                 required: get_option('vaapi-wayland').require(libva.found() and features['wayland']))
```

File: meson.build [hunk 4]
```diff
@@ -1394,36 +1394,38 @@ if features['rpi-mmal']
...
+features += {'vaapi-wayland': vaapi_wayland.found()}
 if features['vaapi-wayland']
     dependencies += vaapi_wayland
 endif
 
-vaapi_x11 = dependency('libva-x11', version: '>= 1.1.0', required: get_option('vaapi-x11').require(features['vaapi']))
-features += {'vaapi-x11': features['vaapi'] and features['x11'] and vaapi_x11.found()}
+vaapi_x11 = dependency('libva-x11', version: '>= 1.1.0', required: get_option('vaapi-x11').require(libva.found() and features['x11']))
```
Comment (Dudemanguy) [inline meson.build:1411]:
```suggestion
vaapi_x11 = dependency('libva-x11', version: '>= 1.1.0',
                                         required: get_option('vaapi-x11').require(libva.found() and features['x11']))
```

**Comment by Dudemanguy** [comment]
Thanks for working on this!
